cloudinit/config/cc_ubuntu_advantage.py

```diff
@@ -288,35 +290,79 @@ def configure_ua(token, enable=None):
         # Allow `ua attach` to fail in already attached machines
         subp.subp(attach_cmd, rcs={0, 2}, logstring=redacted_cmd)
     except subp.ProcessExecutionError as e:
-        error = str(e).replace(token, REDACTED)
-        msg = f"Failure attaching Ubuntu Advantage:\n{error}"
+        err = str(e).replace(token, REDACTED)
+        msg = f"Failure attaching Ubuntu Advantage:\n{err}"
         util.logexc(LOG, msg)
         raise RuntimeError(msg) from e
 
-    enable_errors = []
-    for service in enable:
-        try:
-            cmd = ["ua", "enable", "--assume-yes", service]
-            subp.subp(cmd, capture=True)
-        except subp.ProcessExecutionError as e:
-            if re.search("is already enabled.", str(e)):
-                LOG.debug('Service "%s" already enabled.', service)
-            else:
-                enable_errors.append((service, e))
+    # Enable services
+    if not enable:
+        return
+    cmd = ["ua", "enable", "--assume-yes", "--format", "json", "--"] + enable
+    try:
+        enable_stdout, _ = subp.subp(cmd, capture=True, rcs={0, 1})
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            "Error while enabling service(s): " + ", ".join(enable)
+        ) from e
+
+    try:
+        enable_resp = json.loads(enable_stdout)
+    except json.JSONDecodeError as e:
+        raise RuntimeError(f"UA response was not json: {enable_stdout}") from e
+
+    # At this point we were able to load the json response from UA. This
+    # response contains a list of errors under the key 'errors'. E.g.
+    #
+    #   {
+    #      "errors": [
+    #        {
+    #           "message": "UA Apps: ESM is already enabled ...",
+    #           "message_code": "service-already-enabled",
+    #           "service": "esm-apps",
+    #           "type": "service"
+    #        },
+    #        {
+    #           "message": "Cannot enable unknown service 'asdf' ...",
+    #           "message_code": "invalid-service-or-failure",
+    #           "service": null,
+    #           "type": "system"
+    #        }
+    #      ]
+    #   }
+    #
+    # From our pov there are two type of errors, service and non-service
+    # related. We can distinguish them by checking if `service` is non-null
+    # or null respectively.
+
+    # pylint: disable=import-error
+    from uaclient.messages import ALREADY_ENABLED
+
+    # pylint: enable=import-error
```

Comment:
I added this closing pylint disable/enable comment here. Ultimately, I think we may want to add something like the following to test-requirements.txt:
```diff
diff --git a/test-requirements.txt b/test-requirements.txt
index 739d8ebb4..fb77c867a 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -7,3 +7,6 @@ pytest-mock
 setuptools
 jsonschema
 responses
+
+# Ubuntu-specific support for cc_ubuntu_advantage module in unittests
+uaclient @ git+https://github.com/canonical/ubuntu-advantage-client.git@b5ed2599917423d2633ce6101eb47f17b9f63426
```

But, this needs more discussion on the approach we agree to take as we'd need to think more about impact to non-ubuntu unittest runners.

Currently, the above would break non-ubuntu testing as seen if we were to run `./tools/run-container --package --source-package --unittest centos/7`. But, having the actual library imported for uaclient will provide is with more thorough unittests for Ubuntu and I think eventually e want to get to a unittest path where we don't have to ignore 
pylint warnings on import-error as well as eventually drop the [fake_uaclient fixture](https://github.com/canonical/cloud-init/blob/main/tests/unittests/config/test_cc_ubuntu_advantage.py#L57) for our ubuntu-based testing.   This comment doesn't need to block this current PR though.